src/control/server/storage/config.go

```diff
@@ -241,6 +251,94 @@ func (tcs TierConfigs) Validate() error {
 	return nil
 }
 
+func (tcs TierConfigs) validateBdevTierRoles() error {
+	sc := tcs.ScmConfigs()[0]
+	bcs := tcs.BdevConfigs()
+
+	nrWALTiers := 1
+	nrIndexTiers := 1
+	if sc.Class == ClassDcpm {
+		nrWALTiers = 0
+		nrIndexTiers = 0
+		if len(bcs) > 1 {
+			return errors.Errorf("more than 1 bdev tiers not allowed with dcpm scm class")
+		}
+	}
+
+	var foundWALTiers, foundIndexTiers, foundDataTiers int
+	for _, bc := range bcs {
+		bits := bc.Bdev.DeviceRoles.OptionBits
+		if (bits & BdevRoleWAL) != 0 {
+			foundWALTiers++
+		}
+		if (bits & BdevRoleIndex) != 0 {
+			foundIndexTiers++
+		}
+		if (bits & BdevRoleData) != 0 {
+			foundDataTiers++
+		}
+	}
+
+	if foundWALTiers != nrWALTiers {
+		return errors.Errorf("found %d WAL tiers, wanted %d", foundWALTiers, nrWALTiers)
+	}
+	if foundIndexTiers != nrIndexTiers {
+		return errors.Errorf("found %d Index tiers, wanted %d", foundIndexTiers, nrIndexTiers)
+	}
+	// When bdev NVMe tiers exist, there should always be at least one Data tier.
+	if foundDataTiers == 0 {
+		return errors.New("found 0 Data tiers, wanted at least 1")
+	}
```

Comment:
Hm, if no roles are specified for a tier, should we assume it's data by default?

Reply:
the heuristics to decide what roles to apply is discussed in the ticket